### PR TITLE
#314: complete implementation of upper bound constrained optimizer

### LINKED ISSUE #314: Add upper/lower bounds constraints to optimization algorithms
This is in order to generalize the LB process to include cases where some quantities of interest must remain within pre-determined bounds.

@lifflander @nlslatt

### PR DESCRIPTION
**Pull request by ppebay**
Resolves #314

### CODE REVIEW

**Comment by ppebay** [comment]
Constrained algorithm seems to be working as it should -- but more testing is needed. I will submit this PR nonetheless to get feedback, thanks

Screen Shot 2022-12-07 at 10 36 43 PM


Screen Shot 2022-12-07 at 10 36 49 PM

File: src/lbaf/Execution/lbsTemperedCriterion.py [hunk 1]
```diff
@@ -6,28 +6,38 @@
...
+        # Compute maximum work of original arrangement
         w_max_0 = max(
-            self.work_model.compute(r_src),
-            self.work_model.compute(r_dst))
+            self._work_model.compute(r_src),
+            self._work_model.compute(r_dst))
+
+        # Move objects into proposed new arrangement
+        for o in objects:
+            self._phase.transfer_object(o, r_src, r_dst)
```
Comment (cz4rs) [inline src/lbaf/Execution/lbsTemperedCriterion.py:28]:
```
    File "/home/runner/work/LB-analysis-framework/LB-analysis-framework/src/lbaf/Execution/lbsTemperedCriterion.py", line 28, in compute
      self._phase.transfer_object(o, r_src, r_dst)
  AttributeError: 'TemperedCriterion' object has no attribute '_phase'
```

File: src/lbaf/Execution/lbsTemperedCriterion.py [hunk 2]
```diff
@@ -6,28 +6,38 @@
...
+
+        # Move objects back into original arrangement
+        for o in objects:
+            self._phase.transfer_object(o, r_dst, r_src)
+        
+        # Return criterion value
+        return w_max_0 - w_max_new
 
         # Compute loads in proposed new arrangement
         object_loads = sum([o.get_load() for o in objects])
```
Comment (cz4rs) [inline src/lbaf/Execution/lbsTemperedCriterion.py:43]: `return` here makes the subsequent code unreachable

File: src/lbaf/Execution/lbsTemperedCriterion.py (hunk 1 above)
Comment (cz4rs) [reply src/lbaf/Execution/lbsTemperedCriterion.py:28]: I haven't verified this, but I think `set_phase` is missing somewhere along the way.

File: src/lbaf/Applications/user-defined-memory-toy-problem.yaml [hunk 3]
```diff
@@ -58,7 +58,7 @@ work_model:
     beta: 0.
     gamma: 0.
     upper_bounds:
-      max_memory_usage: 8.0e+9
+      max_memory_usage: 7.0e+9
```
Comment (nlslatt) [inline src/lbaf/Applications/user-defined-memory-toy-problem.yaml:61]: This is not an acceptable change.

File: src/lbaf/Execution/lbsTemperedCriterion.py (hunk 2 above)
Comment (ppebay) [reply src/lbaf/Execution/lbsTemperedCriterion.py]: Yes, it was "scratch code" that I did not want to lose (pure laziness on my part). Thanks

File: src/lbaf/Applications/user-defined-memory-toy-problem.yaml (hunk 3 above)
Comment (ppebay) [reply src/lbaf/Applications/user-defined-memory-toy-problem.yaml]: Thanks, back to 8

**Comment by ppebay** [comment]
If someone gets a chance to re-review this PR @nlslatt @lifflander I'd like to close the implementation branch -- and fan out from there with several follow-on/optimization branches. Thanks!

**Review by nlslatt — APPROVED** [review]

**Pull request merged**